docs/guides/primitive-input-output.ipynb

```diff
@@ -0,0 +1,461 @@
+{
```

Comment:
<p>Aha, I see now that the earlier section was Estimator-focused and this one is Sampler-focused. I think it should then be clarified in the earlier title (Estimator broadcasting rules? or having Estimator output somewhere?). </p><p> </p><p>And here I am going to share what might be just my personal view, feel free to ignore it: I don't think that the DataBin/BitArray-focused explanation is easy to follow for a first-time reader. The information is there, but I think it's "obstructed" by the use of unfamiliar terminology. I would personally focus on the key "actionable" points of the explanation first (which is what I would care about as a user) and then follow-up with the more formal explanation for completeness. These points for me are:</p><p> </p><p>	 There are 2 main differences between the Sampler and Estimator results:</p><p> </p><ol><li>The result <strong>content</strong>: it's measurement data instead of expectation values, the data can be processed into a dictionary with key-value pairs...</li><li>The result <strong>format</strong>:</li></ol><ul><li class="ql-indent-1">the attributes (keys) of the Sampler result data are different to the attributes of the Estimator result data</li><li class="ql-indent-1">In the estimator, if you do <code>result[0].data</code> you will always get the following 3 attributes: <code>evs, stds, ensemble_standard_error</code></li><li class="ql-indent-1">However, in the sampler, if you do <code>result[0].data</code>, there can be more than one attribute with result data, and the name(s) of the attribute(s) containing the measurement data will <strong>change</strong> <strong>depending on the name or names of the classical register(s) in the circuit</strong>. There will be as many attributes as classical registers, and each will match the classical register name. If for example, you have a circuit with two registers called <code>banana</code> and <code>orange</code>, you can access the corresponding measurement data as <code>result[0].data.banana</code> and <code>result[0].data.orange</code>. </li><li class="ql-indent-1">If you don't build the registers manually and just use the <code>QuantumCircuit</code> default/usual construction method, there will always be a single classical register called <code>meas</code>, and the measurement data will be accessed as <code>result[0].data.meas</code>. (Important, this is not because of <code>measure_all</code>, the name of the register depends on how the circuit is <strong>built</strong>).</li></ul><p> </p><p>After this, I would clearly mention the utility methods for processing result data, including:</p><p> </p><ul><li>get_counts() -&gt; gives you counts</li><li>get_bitstrings() -&gt; gives you shot-wise outputs (explain better)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Qiskit/documentation/pull/1621/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Reply:
<p>I am unfamiliar with what an introductory user knows and doesn't know, but I imagine it's pretty varied. Someone coming from ML or tensor networks will be intimately familiar with arrays and broadcasting and be at home with them. Someone coming from an experimental background may not be, though they might also be interested to learn that the values don't have to be BitArray's, they can be IQ data.</p>

Reply:
<p>Independent of the discussion, also posting some more concrete suggestions:</p><p> </p><ol><li>Left a similar comment earlier, but it's important that our docs don't talk about BitArrays as though they're distributions: they're samples from distributions. So "<code>BitArray</code> objects that store the measured probability distribution(s)" -&gt; "BitArray objects that contain samples from the experiment, typically one bitstring per shot."</li></ol><p> </p>